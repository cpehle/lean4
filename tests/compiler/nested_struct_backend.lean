-- Test for nested structs in the struct backend

@[struct]
structure Point where
  x : Float
  y : Float

@[struct] 
structure Rectangle where
  topLeft : Point
  bottomRight : Point

def main : IO Unit := do
  let p1 := Point.mk 1.0 2.0
  let p2 := Point.mk 4.0 6.0
  let rect := Rectangle.mk p1 p2
  
  -- Access nested fields
  let width := rect.bottomRight.x - rect.topLeft.x
  let height := rect.bottomRight.y - rect.topLeft.y
  let area := width * height
  
  IO.println s!"Rectangle: topLeft=({rect.topLeft.x}, {rect.topLeft.y}), bottomRight=({rect.bottomRight.x}, {rect.bottomRight.y})"
  IO.println s!"Width: {width}, Height: {height}"
  IO.println s!"Area: {area}"