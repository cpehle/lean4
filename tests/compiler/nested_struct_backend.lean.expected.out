Rectangle: topLeft=(1.000000, 2.000000), bottomRight=(4.000000, 6.000000)
Width: 3.000000, Height: 4.000000
Area: 12.000000
